pkg/build/registry/buildconfig/webhook.go

```diff
@@ -0,0 +1,75 @@
+package buildconfig
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/generic/rest"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/build/client"
+	"github.com/openshift/origin/pkg/build/webhook"
+)
+
+func NewWebHookREST(registry Registry, instantiator client.BuildConfigInstantiator, plugins map[string]webhook.Plugin) *rest.WebHook {
+	controller := &controller{
+		registry:     registry,
+		instantiator: instantiator,
+		plugins:      plugins,
+	}
+	return rest.NewWebHook(controller, false)
+}
+
+type controller struct {
+	registry     Registry
+	instantiator client.BuildConfigInstantiator
+	plugins      map[string]webhook.Plugin
+}
+
+// ServeHTTP implements rest.HookHandler
+func (c *controller) ServeHTTP(w http.ResponseWriter, req *http.Request, ctx kapi.Context, name, subpath string) error {
+	parts := strings.Split(subpath, "/")
+	if len(parts) < 2 {
+		return errors.NewBadRequest(fmt.Sprintf("unexpected hook subpath %s", subpath))
+	}
+	secret, hookType := parts[0], parts[1]
+	path := ""
+	if len(parts) > 2 {
+		path = strings.Join(parts[2:], "/")
+	}
+
+	plugin, ok := c.plugins[hookType]
+	if !ok {
+		return errors.NewNotFound("BuildConfigHook", hookType)
+	}
+
+	config, err := c.registry.GetBuildConfig(ctx, name)
+	if err != nil {
+		return errors.NewUnauthorized(fmt.Sprintf("the webhook %q for %q did not accept your secret", hookType, name))
+	}
+
+	revision, proceed, err := plugin.Extract(config, secret, path, req)
+	switch err {
+	case webhook.ErrSecretMismatch, webhook.ErrHookNotEnabled:
+		return errors.NewUnauthorized(fmt.Sprintf("the webhook %q for %q did not accept your secret", plugin, name))
+	case nil:
+	default:
+		return errors.NewInternalError(fmt.Errorf("hook failed: %v", err))
+	}
+
+	if !proceed {
+		return nil
```

Comment:
we should log something here, if only at a verbose level.  silent no-ops are the nightmare of production debugging.